src/python/pants/base/specs.py

```diff
@@ -25,6 +26,42 @@ class Spec(AbstractClass):
   def to_spec_string(self):
     """Returns the normalized string representation of this spec."""
 
+  class AddressFamilyResolutionError(Exception): pass
+
+  @abstractmethod
+  def matching_address_families(self, address_families_dict):
+    """Given a dict of (namespace path) -> AddressFamily, return the values matching this spec.
+
+    :raises: :class:`Spec.AddressFamilyResolutionError` if no address families matched this spec.
+    :return: list of AddressFamily.
+    """
+
+  @classmethod
+  def address_families_for_dir(cls, address_families_dict, spec_dir_path):
+    """Implementation of `matching_address_families()` for specs matching at most one directory."""
+    maybe_af = address_families_dict.get(spec_dir_path, None)
+    if maybe_af is None:
+      raise cls.AddressFamilyResolutionError(
+        'Path "{}" does not contain any BUILD files.'
+        .format(spec_dir_path))
+    return [maybe_af]
+
+  class AddressResolutionError(Exception): pass
+
+  def all_address_target_pairs(self, address_families):
```

Comment:
Rather than overriding concrete methods, it would be preferable to extract this as an optional helper as you did with `address_families_for_dir`, that some abstract method can optionally call.

Reply:
This is great! Done in c4d2edc1d569e39cbb885ac47e649c38db3f77c7!